src/types/hooks.ts

```diff
@@ -96,47 +107,54 @@ export type AfterHookLifecycle = 'onAfterRouteEnter' | 'onAfterRouteUpdate' | 'o
  */
 export type HookLifecycle = BeforeHookLifecycle | AfterHookLifecycle
 
-type HookContext<
+type AfterHookContext<
+  TRoute extends Route,
   TRoutes extends Routes,
   TRejections extends Rejections
 > = {
   from: RouterResolvedRouteUnion<TRoutes> | null,
   reject: RouterReject<TRejections>,
   push: RouterPush<TRoutes>,
   replace: RouterReplace<TRoutes>,
+  update: RouteUpdate<ResolvedRoute<TRoute>>,
 }
 
 type BeforeHookContext<
+  TRoute extends Route,
   TRoutes extends Routes,
   TRejections extends Rejections
-> = HookContext<TRoutes, TRejections> & {
+> = {
+  from: RouterResolvedRouteUnion<TRoutes> | null,
+  reject: RouterReject<TRejections>,
+  push: RouterPush<TRoutes>,
+  replace: RouterReplace<TRoutes>,
+  update: RouteUpdate<ResolvedRoute<TRoute>>,
   abort: RouterAbort,
 }
 
 export type BeforeHook<
+  TRoute extends Route = Route,
   TRoutes extends Routes = Routes,
   TRejections extends Rejections = Rejections
-> = (to: RouterResolvedRouteUnion<TRoutes>, context: BeforeHookContext<TRoutes, TRejections>) => MaybePromise<void>
+> = (to: RouterResolvedRouteUnion<TRoutes>, context: BeforeHookContext<TRoute, TRoutes, TRejections>) => MaybePromise<void>
 
 export type AddBeforeHook<
+  TRoute extends Route = Route,
   TRoutes extends Routes = Routes,
   TRejections extends Rejections = Rejections
-> = (hook: BeforeHook<TRoutes, TRejections>) => HookRemove
-
-type AfterHookContext<
-  TRoutes extends Routes = Routes,
-  TRejections extends Rejections = Rejections
-> = HookContext<TRoutes, TRejections>
+> = (hook: BeforeHook<TRoute, TRoutes, TRejections>) => HookRemove
 
 export type AfterHook<
+  TRoute extends Route = Route,
   TRoutes extends Routes = Routes,
   TRejections extends Rejections = Rejections
-> = (to: RouterResolvedRouteUnion<TRoutes>, context: AfterHookContext<TRoutes, TRejections>) => MaybePromise<void>
+> = (to: RouterResolvedRouteUnion<TRoutes>, context: AfterHookContext<TRoute, TRoutes, TRejections>) => MaybePromise<void>
```

Comment:
I'm a bit confused by the simplicity of just adding `TRoute` to before/after hooks. I expected needing both `TTo` and `TFrom` generics. The single route implies that the only route we can accurately narrow is the `from` in context but that's not true right? If the hook is `beforeRouteLeave`, `afterRouteLeave`, and maybe `..Update` this makes sense. But if the hook is `beforeRouteEnter` or `afterRouteEnter` wouldn't `TRoute` actually be the `to`? 

```ts
const myRoute = createRoute({ ... })
myRoute.onBeforeRouteEnter((to, context) => {
  to
// ^? typeof myRoute??
  context.from
// ^? typeof RouterResolvedRouteUnion<TRoutes>??
})
```